future/atomicTransactionComposer.go

```diff
@@ -494,65 +548,204 @@ func (atc *AtomicTransactionComposer) Execute(client *algod.Client, ctx context.
-			log := txinfo.Logs[j]
-			if len(log) >= 4 && bytes.Equal(log[:4], ABI_RETURN_HASH) {
-				failedToFindResult = false
-				abiType, err := abi.TypeOf(txContext.method.Returns.AbiType)
-				if err != nil {
-					returnValues = append(returnValues, ABIResult{TxID: txid, RawReturnValue: log[4:], ReturnValue: nil, DecodeError: err})
-					break
-				}
+		lastLog := methodCallInfo.Logs[len(methodCallInfo.Logs)-1]
+		if !bytes.HasPrefix(lastLog, abiReturnHash) {
+			result.DecodeError = errors.New("method call did not log a return value")
+			returnValues = append(returnValues, result)
+			continue
+		}
+
+		result.RawReturnValue = lastLog[len(abiReturnHash):]
+
+		abiType, err := txContext.method.Returns.GetTypeObject()
+		if err != nil {
+			result.DecodeError = err
+			returnValues = append(returnValues, result)
+			break
+		}
+
+		result.ReturnValue, result.DecodeError = abiType.Decode(result.RawReturnValue)
+		returnValues = append(returnValues, result)
+	}
+
+	return txinfo.ConfirmedRound, atc.getTxIDs(), returnValues, nil
+}
+
+// marshallAbiUint64 converts any value used to represent an ABI "uint64" into
+// a golang uint64
+func marshallAbiUint64(value interface{}) (uint64, error) {
+	abiType, err := abi.TypeOf("uint64")
+	if err != nil {
+		return 0, err
+	}
+	encoded, err := abiType.Encode(value)
+	if err != nil {
+		return 0, err
+	}
+	decoded, err := abiType.Decode(encoded)
+	if err != nil {
+		return 0, err
+	}
+	marshalledValue, ok := decoded.(uint64)
+	if !ok {
+		err = fmt.Errorf("Decoded value is not a uint64")
+	}
+	return marshalledValue, err
+}
+
+// marshallAbiAddress converts any value used to represent an ABI "address" into
+// a golang address string
+func marshallAbiAddress(value interface{}) (string, error) {
+	abiType, err := abi.TypeOf("address")
+	if err != nil {
+		return "", err
+	}
+	encoded, err := abiType.Encode(value)
+	if err != nil {
+		return "", err
+	}
+	decoded, err := abiType.Decode(encoded)
+	if err != nil {
+		return "", err
+	}
+	marshalledValue, ok := decoded.([]byte)
+	if !ok || len(marshalledValue) != len(types.ZeroAddress) {
+		err = fmt.Errorf("Decoded value is not a 32 length byte slice")
+	}
+	var addressValue types.Address
+	copy(addressValue[:], marshalledValue)
+	return addressValue.String(), err
+}
 
-				abiValue, err := abiType.Decode(log[4:])
-				if err != nil {
-					returnValues = append(returnValues, ABIResult{TxID: txid, RawReturnValue: log[4:], ReturnValue: nil, DecodeError: err})
+// populateMethodCallReferenceArgs parses reference argument types and resolves them to an index
+// into the appropriate foreign array. Their placement will be as compact as possible, which means
+// values will be deduplicated and any value that is the sender or the current app will not be added
+// to the foreign array.
+func populateMethodCallReferenceArgs(sender string, currentApp uint64, types []string, values []interface{}, accounts *[]string, apps *[]uint64, assets *[]uint64) ([]int, error) {
```

Comment:
This function is based on https://github.com/algorand/go-algorand/blob/6be3ffa4d2a0499b3c407475338327e758e22132/cmd/goal/application.go#L1091

Reply:
Hoping for generic be introduced to simplify the code 🙏